architecture/crds.md

```diff
@@ -0,0 +1,324 @@
+
+In Kubernetes, the following CRD patterns have emerged:
+
+- Generic CRD: A single CRD "generalizes" a description of something, and any programs that need that generalized object can use that CRD. The CRD is thus a shared resource with no single controller.
+- CRD and Singleton Controller: A specific application defines a specific CRD that only it uses. Only one controller uses that CRD.
+- Common CRD: Two or more tightly coupled programs use a CRD to update information that is shared between the programs. Microservice architectures sometimes use this approach. (ConfigMaps and Secrets are also used this way, and the documentation for CRDs even suggests this as an alternative to ConfigMaps and Secrets in this capacity)
+- CRD and per-namespace controllers: One CRD is used globally, but each namespace that uses this CRD has its own controller which watches only that namespace. This strategy is done to reduce the cluster-level permissions the controller must have.
+
+As a special case, some controllers install and manage their own CRDs. For the most part, we won't deal with these here, as these do not pose a problem for Helm.
+
+The following facts about CRDs should be kept in mind throughout this paper:
+
+- CRDs are mutable. An operator can update versions, schemas, names, etc. on a CRD _ad hoc_. The API server will not enforce restrictions like it does on Deployments or other objects.
+- When a CR (CRD instance) is written to Kubernetes (on an update, for example), it will be _rewritten_ to the version that the CRD has marked as default. This means that backward compatibility can break merely by updating the default version on a CRD object. [See this section of the docs](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#writing-reading-and-updating-versioned-customresourcedefinition-objects), which reveals a few other edge cases.
+    - During a read operation, Kubernetes can rewrite an object version without rewriting the body of the CR. So you can get a resource that is marked with a version that it actually is not.
+    - On a write operation, a version field may be permanently  rewritten to a version other than the version given in the object to be written, but the body is not updated
+    - A retrieved object may not match the schema of the version that is in its apiVersion field because of the above.
+- A developer can install a Kubernetes webhook that will auto-convert CRD fields. The controller does not have visibility into this conversion: It happens before the event triggers inside of Kubernetes.
+
+A quick glance through this section should reveal one stark fact: Everything that uses a CRD should (and perhaps must) use the same _version_ of the CRD as the one marked as the default for that cluster. This is the single most fragile aspect of the entire CRD system.
```

Comment:
An alternative fact to draw here is that the k8s CRD system only works by default (without conversion webhook support) if a CRD is backwards and forwards compatible through its versions, such that the stored data can be consumed under any of the known stored versions, i.e. the `status.storedVersions` field (whose value is going to depend on the cluster).

I'd also suggest that this means that CRD management is in many cases something that must be done by a more-advanced user than the minimal Helm User Role described earlier. This also shows up below that much of the discussion of "Things Helm Users cannot do" are things that a sufficiently-privileged k8s user _can_ do, and since CRDs are global resources, there's already an element of "sufficiently-privileged" (not necessarily the same) needed to work with them at all.

Reply:
Are there k8s bug reports or design discussions relevant to these issues? If so, it'd be useful to reference them here, so we understand how this situation might evolve in the future.

Reply:
> I'd also suggest that this means that CRD management is in many cases something that must be done by a more-advanced user than the minimal Helm User Role described earlier.

I would like it if more advanced users handled CRDs. But, it's not a MUST and people without expertise do in fact install them more often than not. They copy and paste commands to just get their thing to work.

> Are there k8s bug reports or design discussions relevant to these issues? If so, it'd be useful to reference them here, so we understand how this situation might evolve in the future. This document would hopefully also support discussions upstream as to how things can evolve to be better-manageable, if desired.

I'm not sure of any good places that outline this material. SIG API Machinery owns this code (they own like 1/4 of k8s), do not have near enough contributors, have a lot of work to do, and don't have good solutions for this the last I checked. If you are interested in helping out there than that SIG is the one to get involved with.

Reply:
I should clarify that I was thinking of 'CRD management in many cases' in the sense of "everything except installs new CRDs", i.e. the things that Helm v3 doesn't provide mechanisms for yet. I don't currently believe that (beyond particular provably-safe cases I've described elsewhere but not had time to try and _demonstrate_), the bare-minimum Helm user expertise level can by definition fully-correctly/safely manage CRDs after installation. for the reasons described in this document, without either narrower restrictions on CRD modifications, or some intrusive behaviour changes around CRD version immutability or storage format behaviours, to turn some of the fragility into early hard-failures irrespective of the tool used.

That's why I was hoping someone had already written this stuff up, and narrowed down good ideas and rejected bad ideas already. ^_^ I doubt I'll get a chance to chase it that far, but I might poke around SIG API Machinery as a general familiarisation. Thank you for the pointer.

Reply:
API Machinery member here. I've read this document and the initial issue about updating CRDs. Thanks @mattfarina for mentioning us. We are very busy but we'd love to help and collaborate. I don't know if we have a specific solution to the problem here yet, it's obviously not an easy problem to solve, but please feel free to reach out early when you have valuable feedback like this!